### DISCUSSION TITLE
Sync timer for VideoWriter to a keypress

### DISCUSSION THREAD

**Discussion by bruno-f-cruz** [bonsai-rx/bonsai#868] [Q&A] [answered]
The following question was asked in the Bonsai Users - Google Groups
[Link to original post](https://groups.google.com/g/bonsai-users/c/ms6ByAKp_JQ)

---
**Original post**

Hi all, simple question. I'm looking to create a pipeline that has independent control of VideoWriter from VideoDeviceCapture, so I can preview the assay from VideoDeviceCapture without it recording, and then begin recording with a keypress (like this question here https://groups.google.com/g/bonsai-users/c/gjOneBsc_gQ/m/1pyoAXbCBQAJ). I need to record for a specified amount of time, which is pretty simple to implement from this question here (https://groups.google.com/g/bonsai-users/c/SGlIshGtnMw/m/UU_wf1KxAwAJ), however I am running into an issue where the timer automatically starts when I begin the pipeline, rather than when I perform the keypress, and thus the workflow ends automatically without me ever recording in the first place. When I swap out SkipUntil for SubscribeWhen, however, I don't see the camera preview. Does anybody know how to sync my timer to begin when I perform the keypress? Thank you for help, my very simple pipeline is attached below. 

---

**Comment by bruno-f-cruz** [comment, marked as answer]
See a possible solution attached.
In this workflow, the camera stream is constantly "on" by publishing it to a subject (`Camera`). For each key press (F1), we trigger a  `SelectMany` that immediately subscribes to the camera stream and starts saving data. In parallel we start a `Timer` and set its `DueTime` property. Once this last duration elapses, the timer will close the `TakeUntil` and subsequently close the `SelectMany`. 

Let me know if it works :)

[RecordTriggeredVideo.zip](https://github.com/bonsai-rx/bonsai/files/8440851/RecordTriggeredVideo.zip)

**Reply by chiyu1203** [reply]
Hi Bruno, I adopted this workflow and added tracking method into it. However, it does not behave what I was expected. I would like to the camera stream and tracking on right after I started the workflow, so that the background subtraction can be done before introducing the animal. Then I place the animal and start video recording + receiving tracking information + the timer with KeyboardPress .
However, it looks like only the camera stream is turned on when I starts the workflow (the tracking stream only starts after the keyboard presses). Is there anything I did wrong with the tracking stream?
[preference_test_circular_timer4.zip](https://github.com/bonsai-rx/bonsai/files/11882075/preference_test_circular_timer4.zip)

**Reply by chiyu1203** [reply]
I found the bug in my workflow! Problem solved now! Your answer is great! But I have came up another question about publish, replay, resource subject in SelectMany. I will create a new question to address my concern.
